exotations/solvers/exotica_ddp_solver/src/analytic_ddp_solver.cpp

```diff
@@ -0,0 +1,112 @@
+namespace exotica
+{
+AnalyticDDPSolver::AnalyticDDPSolver() = default;
+AnalyticDDPSolver::~AnalyticDDPSolver() = default;
+
+void AnalyticDDPSolver::Instantiate(const AnalyticDDPSolverInitializer& init)
+{
+    parameters_ = init;
+    base_parameters_ = parameters_;
+}
+
+void AnalyticDDPSolver::BackwardPass()
+{
+    constexpr double min_clamp_ = -1e10;
+    constexpr double max_clamp_ = 1e10;
+    const int T = prob_->get_T();
+    const int NX = prob_->get_num_positions();
+    const int NU = prob_->get_num_controls();
+    const int N = NX + prob_->get_num_velocities();
+    const double dt_squared = dynamics_solver_->get_dt() * dynamics_solver_->get_dt();
+
+    Eigen::MatrixXd Qx, Qu, Qxx, Quu, Qux, Quu_inv, Vxx;
+    Eigen::VectorXd Vx;
+
+    // V = prob_->GetStateCost(T - 1);
+    Vx = prob_->GetStateCostJacobian(T - 1);
+    Vxx = prob_->GetStateCostHessian(T - 1);
+
+    // concatenation axis for tensor products
+    Eigen::array<Eigen::IndexPair<int>, 1> dims = {Eigen::IndexPair<int>(1, 0)};
+
+    for (int t = T - 2; t > 0; t--)
+    {
+        Eigen::VectorXd x = prob_->get_X(t), u = prob_->get_U(t);
+        Eigen::MatrixXd fx = dynamics_solver_->fx(x, u),
+                        fu = dynamics_solver_->fu(x, u);
+
+        fx = fx * dynamics_solver_->get_dt() + Eigen::MatrixXd::Identity(fx.rows(), fx.cols());
+        fu = fu * dynamics_solver_->get_dt();
+
+        // Q = prob_->GetStateCost(t) + prob_->GetControlCost(t) + V; // l + v
+        Qx = prob_->GetStateCostJacobian(t) + fx.transpose() * Vx;  // lx + fx @ Vx
+        Qu = prob_->GetControlCostJacobian(t) + fu.transpose() * Vx;
+        Qxx = prob_->GetStateCostHessian(t) + fx.transpose() * Vxx * fx;
+        Quu = prob_->GetControlCostHessian() + fu.transpose() * Vxx * fu;
+
+        // NOTE: Qux = Qxu for all robotics systems I have seen
+        //  this might need to be changed later on
+        // WARNING_NAMED("ControlLimitedDDPSolver", "Assuming fux = fxu. Make sure this is valid for your robot!");
+        Qux = prob_->GetStateControlCostHessian() + fu.transpose() * Vxx * fx;
+
+        if (parameters_.UseSecondOrderDynamics)
+        {
+            Eigen::Tensor<double, 1> Vx_tensor = Eigen::MatrixToTensor((Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>&)Vx, N);
+            Qxx += Eigen::TensorToMatrix((Eigen::Tensor<double, 2>)dynamics_solver_->fxx(x, u).contract(Vx_tensor, dims),
+                                         N, N) *
+                   dt_squared;
+            Quu += Eigen::TensorToMatrix((Eigen::Tensor<double, 2>)dynamics_solver_->fuu(x, u).contract(Vx_tensor, dims),
+                                         NU, NU) *
+                   dt_squared;
+            Qux += Eigen::TensorToMatrix((Eigen::Tensor<double, 2>)dynamics_solver_->fxu(x, u).contract(Vx_tensor, dims),
+                                         NU, N) *
+                   dt_squared;
+        }
+
+        Quu_inv = (Eigen::MatrixXd::Identity(Quu.rows(), Quu.cols()) *
```

Comment:
Why are these each on their own lines? Somehow the formatting got totally messed up. If you have a preferred format yourself, just turn off clang ``// clang-format off`` and then on again ``// clang-format on``.